File: client/src/main/java/io/xol/chunkstories/sound/SoundsLibrary.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.sound;

import java.util.HashMap;
import java.util.Map;

import io.xol.chunkstories.client.ClientImplementation;
import io.xol.chunkstories.sound.ogg.SoundDataOggSample;
import io.xol.chunkstories.sound.ogg.SoundDataOggStream;

public class SoundsLibrary {
	private final ClientImplementation client;
	static Map<String, SoundData> soundsData = new HashMap<>();

	public SoundsLibrary(ClientImplementation client) {
		this.client = client;
	}

	public SoundData obtainOggSample(String soundEffect) {
		SoundDataOggSample sd = new SoundDataOggSample(client.getContent().getAsset(soundEffect));
		sd.name = soundEffect;
		if (sd.loadedOk())
			return sd;
		return null;
	}

	public SoundData obtainSample(String soundEffect) {
		if (soundEffect == null)
			return null;

		if (!soundEffect.startsWith("./"))
			soundEffect = "./" + soundEffect;

		SoundData sd = soundsData.get(soundEffect);
		if (sd != null)
			return sd;
		if (soundEffect.endsWith(".ogg")) {
			sd = obtainOggSample(soundEffect);
		} else
			return null;
		if (sd == null)
			return null;
		soundsData.put(soundEffect, sd);
		return sd;
	}

	public void clean() {
		for (SoundData sd : soundsData.values()) {
			sd.destroy();
		}
		soundsData.clear();
	}

	/**
	 * This methods returns an unique SoundData used exclusivly for a specific
	 * SoundSourceBuffered
	 * 
	 * @param musicName
	 * @return
	 */
	public SoundData obtainBufferedSample(String musicName) {
		if (!musicName.startsWith("./"))
			musicName = "./" + musicName;

		SoundDataBuffered sd;
		if (musicName.endsWith(".ogg")) {
			sd = obtainOggStream(musicName);
		} else
			return null;
		return sd;
	}

	private SoundDataBuffered obtainOggStream(String musicName) {
		SoundDataOggStream sd;

		sd = new SoundDataOggStream(client.getContent().getAsset(musicName).read());
		sd.name = musicName;
		if (sd.loadedOk())
			return sd;

		return null;
	}

}

File: client/src/main/java/xyz/chunkstories/sound/SoundsLibrary.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.sound

import java.util.HashMap

import xyz.chunkstories.client.ClientImplementation
import xyz.chunkstories.sound.ogg.SoundDataOggSample
import xyz.chunkstories.sound.ogg.SoundDataOggStream

class SoundsLibrary(private val client: ClientImplementation) {

    fun obtainSample(soundEffect: String?): SoundData? {
        if (soundEffect == null)
            return null

        return soundsData.getOrPut(soundEffect) {
            if (soundEffect.endsWith(".ogg")) {
               loadOggSample(soundEffect)
            } else throw Exception("Unhandled extension: $soundEffect")
        }
    }

    private fun loadOggSample(sampleName: String): SoundData {
        val sampleData = SoundDataOggSample(client.content.getAsset(sampleName))
        sampleData.name = sampleName
        return if (sampleData.loadedOk()) sampleData else throw Exception("Failed to load sample $sampleName")
    }

    fun obtainBufferedSample(musicName: String): SoundDataBuffered? {
        val soundDataBuffered: SoundDataBuffered?
        when {
            musicName.endsWith(".ogg") -> soundDataBuffered = obtainOggStream(musicName)
            else -> return null
        }
        return soundDataBuffered
    }

    private fun obtainOggStream(streamedSampleName: String): SoundDataBuffered? {
        val soundData = SoundDataOggStream(client.content.getAsset(streamedSampleName)!!.read())
        soundData.name = streamedSampleName
        return if (soundData.loadedOk()) soundData else throw Exception("Failed to open stream $streamedSampleName")

    }

    fun cleanup() {
        for (soundData in soundsData.values) {
            soundData.destroy()
        }
        soundsData.clear()
    }

    companion object {
        private val soundsData = HashMap<String, SoundData>()
    }
}
